anaDACScan.py

```diff
@@ -0,0 +1,288 @@
+            second = line.split(' ')[1].strip(' ')
+            tuple_calInfo = parseCalFile(second)
+            calInfo[int(first)] = {'slope' : tuple_calInfo[0], 'intercept' : tuple_calInfo[1]}
+
+    #for each OH, check if calibration files were provided, if not search for the calFile in the $DATA_PATH, if it is not there, then skip that OH for the rest of the script
+    for oh in ohArray:
+        if oh not in calInfo.keys():
+            calDacCalFile = "{0}/{1}/calFile_calDac_{1}.txt".format(dataPath,chamber_config[oh])
+            calDacCalFileExists = os.path.isfile(calDacCalFile)
+            if not calDacCalFileExists:
+                print("Skipping OH{0}, detector {1}, missing CFG_CAL_DAC Calibration file:\n\t{2}".format(
+                    oh,
+                    chamber_config[oh],
+                    calDacCalFile))
+                ohArray = np.delete(ohArray,(oh))
+
+    if len(ohArray) == 0:
+        print('No OHs with a calFile, exiting.')
+        exit(1)
+           
+    # Initialize nested dictionaries
+    dict_DACvsADC_Graphs = nesteddict()
+    dict_RawADCvsDAC_Graphs = nesteddict()
+    dict_DACvsADC_Funcs = nesteddict()
+
+    # Initialize a TGraphErrors for each vfat
+    for oh in ohArray:
+        for vfat in range(0,24):
+             dict_DACvsADC_Graphs[oh][vfat] = r.TGraphErrors()
+             dict_RawADCvsDAC_Graphs[oh][vfat] = r.TGraphErrors()
+             #the reversal of x and y is intended - we want to plot the nameX variable on the y-axis and the nameY variable on the x-axis
+             dict_DACvsADC_Graphs[oh][vfat].GetXaxis().SetTitle(nameY)
+             dict_DACvsADC_Graphs[oh][vfat].GetYaxis().SetTitle(nameX)
+             dict_RawADCvsDAC_Graphs[oh][vfat].GetXaxis().SetTitle(nameY)
+             dict_RawADCvsDAC_Graphs[oh][vfat].GetYaxis().SetTitle(nameX)
+
+    outputFiles = {}         
+             
+    for oh in ohArray:
+        if scandate == 'noscandate':
+            outputFiles[oh] = r.TFile(elogPath+"/"+chamber_config[oh]+"/"+args.outfilename,'recreate')
+        else:    
+            outputFiles[oh] = r.TFile(dataPath+"/"+chamber_config[oh]+"/dacScans/"+scandate+"/"+args.outfilename,'recreate')
+
+    # Loop over events in the tree and fill plots
+    for event in dacScanFile.dacScanTree:
+        oh = event.link
+        vfat = event.vfatN
+
+        calibrated_ADC_value=calInfo[oh]['slope'][vfat]*event.dacValY+calInfo[oh]['intercept'][vfat]
+        calibrated_ADC_error=calInfo[oh]['slope'][vfat]*event.dacValY_Err+calInfo[oh]['intercept'][vfat]
+
+        #Use Ohm's law to convert the currents to voltages. The VFAT3 team told us that a 20k ohm resistor was used.
+        if nominalDacValues[nameX][1][1] == "A":
+
+            calibrated_ADC_value = calibrated_ADC_value/20000.0
+            calibrated_ADC_error = calibrated_ADC_error/20000.0
+
+            if nameX != 'CFG_IREF':
+                calibrated_ADC_value -= nominal_iref 
```

Comment:
I think you have a unit mismatch here.  You're assuming `calibrated_ADC_*` will be in volts (the calibration coef's I suspect are in mV) and so you're taking `mV / 20kOhm` and  assigning this to a current.  But then you're subtracting `nominal_iref` from it but the units on `nominal_iref` are not guaranteed to match the units on `calibrated_ADC_*`.  I think you need to double check this and implement a check in the code to ensure it comes out correct.  

This should also be around line 250 where you set `dict_dacVals`.  You should check the requested DAC value doesn't go outside the maximum range (otherwise it will roll over, e.g. 256 in an 8-bit register will be written as `0x1`.

I think you can check against [maxVfat3DacSize](https://github.com/cms-gem-daq-project/cmsgemos/blob/61538ef0b357202cef902edfd4169f81d1f25bc9/gempython/tools/amc_user_functions_xhal.py#L14).

If greater than 255 is returned it should at the bare minimum throw a warning of some kind.

Reply:
ok, the conversion to V from mV was done in commit d52bfa5b505d2ec8a1cacde0d980e9f777d97f29

I would need to update my gempython to get the function maxVfat3DacSize. Do you recommend I do that? Or I can just use 255 as the maximum value for each register.

Reply:
> ok, the conversion to V from mV was done in commit d52bfa5

That wasn't what I was really getting at.  Say someone is scanning `CFG_BIAS_PRE_I_BIT`.  The current here will be this bias current plus the IREF current.  But looking at lines 124-148 you've correctly enabled `nominal_iref` to be changed from mA, uA, or nA.  But the ADC will *always* read in the voltage unit it's designed to deliver in.  So you could get in a situation where `nominal_iref` is in `nA` but the reading coming from the ADC is converted into `uA` then say 10000 nA subtracted from 32 uA will in this case result in a number that is -9968.  

I think it's better to just stick with `uA` and `mV` inform the user these are the units expected, and convert all values to these (especially below in `nominalDacValues`).

> I would need to update my gempython to get the function maxVfat3DacSize. Do you recommend I do that? Or I can just use 255 as the maximum value for each register.

Don't do this, not every register is an 8 bit register!!  This will cause exactly what we are trying to avoid, rollover.

Build [cmsgemos_gempython](https://mattermost.web.cern.ch/cms-gem-daq/pl/3woxpxxsfpb4tk571bt3x5bw1h)

Reply:
I have implemented an upper and a lower check of the nominal DAC value now.

Reply:
> I have implemented an upper and a lower check of the nominal DAC value now.

I see that, but you didn't really address my main concern.  The units here may mismatch and you haven't made any check against that.  There should be some check which *prevents* subtract a number in `nA` from a number in `uA`.

Reply:
The easiest way to address this would be:

> I think it's better to just stick with uA and mV inform the user these are the units expected, and convert all values to these (especially below in nominalDacValues).

You could inform them in the help menu.

Reply:
sorry, I don't understand what the problem is or what you would like me to do

> So you could get in a situation where `nominal_iref` is in `nA` but the reading coming from the ADC is converted into `uA` then say 10000 nA subtracted from 32 uA will in this case result in a number that is -9968.

32 uA - 10000 nA = 22 uA, and this is correctly done in my script, I convert everything to amperes before doing the subtraction

There are two registers for which the nominal current is in nA in the VFAT3 manual and the rest are in uA or mV in the VFAT3 manual. I can just remove those two registers from the table or exit(1) if the user tries to use the script with those two registers. That seems to be what is implied from your comments. Could you please clarify?

Reply:
> I convert everything to amperes before doing the subtraction

Okay it was not clear to me, but this is good...guess I've been starting at a screen for too long today.